File: 74-search-a-2d-matrix/search-a-2d-matrix.ts
function searchMatrix(matrix: number[][], target: number): boolean {
    let r = matrix.length;
    let c = matrix[0].length;
    let l = 0;
    let h = r * c - 1;

    const getRow = (num: number): number => Math.floor(num / c);
    const getCol = (num: number): number => num % c;

    while (l <= h) {
        let m = Math.floor((l + h) / 2);
        let val = matrix[getRow(m)][getCol(m)]
        if (target === val) return true;
        else if (val < target) l = m + 1;
        else h = m - 1;
    }
    return false;
};

File: 74-search-a-2d-matrix/search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int row = matrix.length;
        int col = matrix[0].length;
        int low = 0;
        int high = row * col - 1;
        while(low <= high) {
            int mid = low + (high - low) / 2;
            int curRow = mid / col;
            int curCol = mid % col;
            int curVal = matrix[curRow][curCol];
            if( curVal == target) return true;
            else if (curVal < target) low = mid + 1;
            else high = mid - 1;
        }
        return false;
    }
}